### DISCUSSION TITLE
macOS builtins.MemoryError

### DISCUSSION THREAD

**Discussion by thedandroid** [evennia/evennia#2637] [Evennia Questions] [answered]
Hey there,

I've been trying, unsuccessfully, to get Evennia running on macOS 12.2. I'm using pyenv and pyenv-virtualenv to manage python versions and virtual environments. I've tried with various versions of Python from 3.8.9 to 3.9.10 all with the same traceback once the server tries to start. I know that this was an issue in the past with a version of python2 and Xcode. I've updated Xcode to 13.2.1. I've tried on two MacBooks with the same result.

```
File "/Users/dallen/.pyenv/versions/3.9.10/envs/mud/lib/python3.9/site-packages/django/db/backends/sqlite3/base.py", line 423, in execute
         return Database.Cursor.execute(self, query, params)
builtins.MemoryError:
```

Any ideas what could be causing the issue?

**Comment by ghost** [comment]
The only way that I was able to get Evennia to work on macOS, I had to use MacPorts to install alternatives to the system Python and sqlite. This was ages ago (was on Catalina at the time) but it's probably the same permissions issues.

**Reply by thedandroid** [reply]
I'm using a virtual environment with a more up-to-date version of Python, it shouldn't be seeing the system Python at all.

**Comment by jaborsh** [comment]
> Hey there,
> 
> I've been trying, unsuccessfully, to get Evennia running on macOS 12.2. I'm using pyenv and pyenv-virtualenv to manage python versions and virtual environments. I've tried with various versions of Python from 3.8.9 to 3.9.10 all with the same traceback once the server tries to start. I know that this was an issue in the past with a version of python2 and Xcode. I've updated Xcode to 13.2.1. I've tried on two MacBooks with the same result.
> 
> ```
> File "/Users/dallen/.pyenv/versions/3.9.10/envs/mud/lib/python3.9/site-packages/django/db/backends/sqlite3/base.py", line 423, in execute
>          return Database.Cursor.execute(self, query, params)
> builtins.MemoryError:
> ```
> 
> Any ideas what could be causing the issue?

Your Traceback states that you're encountering a Memory Error. Make sure you're running a 64-bit version of Python. 32-bit programs are limited to 4GB of RAM.

**Reply by thedandroid** [reply]
I can confirm that I am using a 64-bit version of Python 3.9.10 currently (I've tried many different versions of Python, however, but all of them 64-bit).

```
>>> import platform
>>> print(platform.architecture())
('64bit', '')
```

**Comment by davidrideout** [comment]
I ran into the same issue a few weeks ago on two macbooks and multiple python versions. 

My workaround was to use a plain `virtualenv` (no `pyenv`).

**Reply by thedandroid** [reply]
I gave it a go and, unfortunately, got the same result ```builtins.MemoryError```

**Comment by ChrisLR** [comment]
I had a similar error awhile back and it was related to permissions on the SQLITE database file.
I chmodded it to 775 (since it was local) and it worked

**Reply by ChrisLR** [reply]
maybe try a chown too..

**Reply by thedandroid** [reply]
Unfortunately, this doesn't seem to be the issue either, thanks for the idea

**Comment by atebyagrue** [comment]
I was able to run it on OSX, but debugging with attach to process in Pycharm was non-existent. I pursued a bit before finally cashing out and just working on Windows. On OSX, I had it running in virtualenv.

**Comment by Griatch** [comment]
Alas, I don't have a Mac to test with; it's strange that Mac would cause an issue like this. Master-branch Evennia supports python 3.10 since a few days though, so try updating to the latest and see if you have more luck with Python 3.10 :confused: 
For a memory error to manifest, I guess one possibility is that the issue lies with a C-based module like Twisted; might be worth experimenting with older/newer Twisteds (or to try with `develop` branch` which ups the dependencies a lot to see if that helps.

**Comment by audiodude** [comment, marked as answer]
For anyone still following this thread (I found it in a google search), see #2854.

I resolved the issue as follows:

1. Install homebrew (`brew`)
1. `$ brew install sqlite` (note, this installs sqlite3 in so-called "keg-only" mode, meaning it is not the default sqlite3 on your system)
1. Install `asdf` (https://asdf-vm.com/) and the asdf python plugin (`$ asdf plugin-add python`)
1. Install a version of python 3.9 (doesn't have to be 3.9.13 exactly) with a link to your brew installed sqlite: `$ CFLAGS="-I$(brew --prefix sqlite)/include"  LDFLAGS="-L$(brew --prefix sqlite)/lib" asdf install python 3.9.13`
1. Create a venv with your new asdf python and install Evennia (`asdf global python 3.9.13` or `asdf local python 3.9.13` then `python -m venv venv` and `pip install -e <path_to_evennia>`)
1. Use Evennia as normal

These are the exact steps I used on macos 10.15.7, where my system installed sqlite reported:

```
$ sqlite3 --version
3.28.0 2019-04-15 14:49:49 378230ae7f4b721c8b8d83c8ceb891449685cd23b1702a57841f1be40b5daapl
```

but homebrew reported:

```
$ brew info sqlite
==> sqlite: stable 3.39.3 (bottled) [keg-only]
```

The same general concept will work on other systems (pyenv especially since asdf uses the same python-build subsystem as pyenv), so long as you provide the CFLAGS= and LDFLAGS= environment variables to your python build tool, so that it picks up the homebrew sqlite instead of the system one.

Sadly I don't have the expertise or patience to trace down the specific bug in sqlite 3.28.0 that is root causing this bug. Hope this helps!
